### PR TITLE
Add Oracle Linux support

### PR DESCRIPTION
**Pull request by Mr-Tao**

### CODE REVIEW

File: eng/common/native/init-distro-rid.sh [hunk 1]
```diff
@@ -20,7 +20,7 @@ getNonPortableDistroRid()
         # shellcheck disable=SC1091
         if [ -e "${rootfsDir}/etc/os-release" ]; then
             . "${rootfsDir}/etc/os-release"
-            if [ "${ID}" = "rhel" ] || [ "${ID}" = "rocky" ] || [ "${ID}" = "alpine" ]; then
+            if [ "${ID}" = "rhel" ] || [ "${ID}" = "rocky" ] || [ "${ID}" = "alpine" || "${ID}" == "ol" ]; then
```
Comment (filipnavara) [inline eng/common/native/init-distro-rid.sh:23]: s/` || `/`] || [`/ ? (should match the others)

Comment (filipnavara) [reply eng/common/native/init-distro-rid.sh:23]: (and single `=`)

Comment (teo-tsirpanis) [reply eng/common/native/init-distro-rid.sh:23]:
```suggestion
            if [ "${ID}" = "rhel" ] || [ "${ID}" = "rocky" ] || [ "${ID}" = "alpine" ] || [ "${ID}" == "ol" ]; then
```

File: src/native/corehost/hostmisc/pal.unix.cpp [hunk 2]
```diff
@@ -767,6 +767,7 @@ pal::string_t normalize_linux_rid(pal::string_t rid)
     pal::string_t rhelPrefix(_X("rhel."));
     pal::string_t alpinePrefix(_X("alpine."));
     pal::string_t rockyPrefix(_X("rocky."));
+    pal::string_t olPrefix(_X("ol."));
```
Comment (jkoritzinsky) [inline src/native/corehost/hostmisc/pal.unix.cpp:770]: @richlander @elinor-fung now that we've frozen the RID graph, we shouldn't need these changes, right?

Comment (am11) [reply src/native/corehost/hostmisc/pal.unix.cpp:770]: Correct.  Host should not care about distro-specific RIDs. This implementation is for legacy .NET versions.

**Comment by am11** [comment]
Duplicate of https://github.com/dotnet/runtime/pull/81115#issuecomment-1608168803. As @agocke wrote, we are not going to feed the entire database of (600+) linux distros in runtime. The new way is to use source build with your preferred RID tag without changing the frozen code.

**Comment by Mr-Tao** [comment]
> Duplicate of [#81115 (comment)](https://github.com/dotnet/runtime/pull/81115#issuecomment-1608168803). As @agocke wrote, we are not going to feed the entire database of (600+) linux distros in runtime. The new way is to use source build with your preferred RID tag without changing the frozen code.

This change does not add any RIDs. It is the barest minimum required to build runtime on Oracle Linux.

**Comment by am11** [comment]
We don't want to add distros related info anywhere in the repo. You can build it just fine with source-build using whatever RID you please. As far as the runtime is concerned, it is just linux.

**Comment by Mr-Tao** [comment]
Should I create a PR against individual release/x branches instead? There are ol RIDs in all dotnet releases, but the support is broken, so they do not build.

**Comment by richlander** [comment]
Are you trying update/fix all (supported) .NET versions or just .NET 8?

I know that the host has some logic for specific distros. It looks like this is what this PR is changing. I cannot remember what our final plan was for that with .NET 8.

@elinor-fung @omajid

**Comment by omajid** [comment]
@tmds do you recall the exact 8.0 behaviour for non-portable RIDs?

**Comment by tmds** [comment]
For .NET 8, non-portable rids are used to identify a source-built .NET made on a distro.
The rid gets baked into the SDK and host at build time.

The vmr uses the `init-distro-rid.sh` script to determine the default non-portable rid for the distro.
You may want to change it in the dotnet/arcade repo, and then it will get synced to the other repos like this one.

**Comment by richlander** [comment]
I think the specific question was related to the RID roll-forward behavior for OSes that have compatible dot releases, like RHEL. AFAIK, this behavior is/was in the host.

**Comment by tmds** [comment]
> I think the specific question was related to the RID roll-forward behavior for OSes that have compatible dot releases, like RHEL. AFAIK, this behavior is/was in the host.

Part of the simplifications we made to the rid-handing in .NET 8 is based on no longer supporting picking between different assets identified by non-portable rids.

This means we also no longer need to express the compatibility between different non-portable rids.

**Comment by richlander** [comment]
Got it. That sounds like a good approach for the Oracle ecosystem. Yes?

**Comment by am11** [comment]
We can simply pass the TargetRid from VMR without continuously updating these frozen parts. This is a script for Oracle Linux:

```sh
$ git clone https://github.com/dotnet/dotnet vmr --single-branch --depth 1 --branch v8.0.3
$ docker run --rm -w /vmr -v$(pwd)/vmr:/vmr amd64/oraclelinux:9 sh -c \
  'dnf install --setopt tsflags=nodocs --refresh -y  automake clang cmake curl findutils git krb5-devel libicu-devel lldb-devel llvm lttng-ust openssl-devel python3 tar zlib-devel;
   dnf --enablerepo=ol9_codeready_builder install lttng-ust-devel;
   ./prep.sh;
   ./build.sh --clean-while-building -- -p:TargetRid=ol.9.1-x64'
```

took ~65 mins on GitHub Actions (with `runs-on: ubuntu-latest`)

**Comment by tmds** [comment]
> We can simply pass the TargetRid

Yes, the rid is just a name, and it can be set using this property.

@Mr-Tao I don't think anyone will object if you add Oracle Linux to this line: https://github.com/dotnet/arcade/blob/4e73daf131ef02e8264b0d3e850c4bdafe6e02b6/eng/common/native/init-distro-rid.sh#L23.

When that rippled down, all the repos will consistently name it using `ol.<major>` instead of `ol.<major>.<minor>`.

If you want the change in earlier .NET versions, I think you should apply the patches in your build.
- For .NET 8, the patch is trivial (change the vmr `init-distro-rid.sh` line).
- I don't think it's worth looking into for .NET 7 (EOL in May).
- For .NET 6, if you think it's worth it (EOL in November), applying the patches will spare upstream maintainers from reviews.

I'll close this PR. If there's a change, it should be in arcade `init-distro-rid.sh`.

**Comment by tmds** [comment]
@richlander @omajid to summarize earlier discussion about non-portable rids in .NET 8+:

_A non-portable rid is a namespace for a distro's source-built .NET assets._

The other use-cases we had pre-.NET 8 are "legacy".

**Comment by Mr-Tao** [comment]
@richlander, I want to fix all supported versions. I’ve started with the main because, in the past, I was asked to commit to the main before making a PR for servicing.

@tmds  We follow RHEL and have already dropped RIDs for minor releases internally.
OL support in .NET 6 official releases is insufficient for people to build it independently.
Given the cadence of releases, it is worth fixing it. I will file a PR for servicing.

**Comment by tmds** [comment]
> Maybe we should make it a required parameter in VMR's main build script and remove this auto distro detection altogether.

Until some change like this happens, I think we should treat it the same as Rocky and RHEL in the script.
There is little effort in adding it. PR is https://github.com/dotnet/arcade/pull/14595.

> Given the cadence of releases, it is worth fixing it. I will file a PR for servicing.

I'm not sure what versions you mean. Depending on the work needed on your end, you may want to check up-front if the maintainers will accept the change for those versions.

**Comment by Mr-Tao** [comment]
> I'm not sure what versions you mean.
I’m referencing the .NET 6 errata Red Hat is issuing frequently.

**Comment by tmds** [comment]
ack, .NET 6.

> We follow RHEL and have already dropped RIDs for minor releases internally.
OL support in .NET 6 official releases is insufficient for people to build it independently.

Are you trying to start building .NET 6 as an Oracle Linux distro maintainer at this time?
Or do you have this working, and is it for the Oracle Linux users to be able to build themselves?

I assume for the latter, the Oracle Linux users should be able to start from the source rpms used by the Oracle Linux distro maintainers?

For the first, either patches go into rpms or they go upstream.

I don't think upstream has a particular interest in these patches, so depending on the work involved, you may want to check up-front if the maintainers will accept them.

**Pull request closed without merging**